File: C++/count-negative-numbers-in-a-sorted-matrix.cpp
class Solution {
public:
    int countNegatives(vector<vector<int>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        int count = 0;
        int i = m - 1;
        int j = 0;
        while ((i >= 0) && (j < n)) {
            if (grid[i][j] < 0) {
                i--;
                count += n - j;
            } else {
                j++;
            }            
        }
        return count;
    }
};

File: TypeScript/count-negative-numbers-in-a-sorted-matrix.ts
function countNegatives(grid: number[][]): number {
    const m = grid.length
    const n = grid[0].length
    let count = 0
    for (let row = 0; row < m; row++) {
        //console.log(grid[row])
        let left = 0, right = n - 1
        if (grid[row][left] < 0) {
            count += n
        } else if (grid[row][right] >= 0) {
            // no negative numbers in this row
        } else {
            while (left < right) {
                const col = Math.floor((left + right) / 2)
                if (grid[row][col] >= 0) {
                    left = col + 1
                } else {
                    right = col
                }
            }
            //console.log("First negative index is ", right, "value", grid[row][right], "count of negatives", n - right)
            count += n - right
        }
    }
    return count
};